otter/test/convergence/test_gathering.py

```diff
@@ -213,85 +221,87 @@ class GetScalingGroupServersTests(SynchronousTestCase):
     """
 
     def setUp(self):
-        """Save basic reused data."""
-        self.req = (ServiceType.CLOUD_SERVERS, 'GET',
-                    'servers/detail?limit=100')
-
-    def test_with_changes_since(self):
-        """
-        If given, servers are fetched based on changes_since
-        """
-        since = datetime(2010, 10, 10, 10, 10, 0)
-        eff = resolve_retry_stubs(
-            get_scaling_group_servers(changes_since=since))
-        fake_response = object()
-        body = {'servers': []}
-        result = resolve_svcreq(
-            eff, (fake_response, body),
-            *svc_request_args(changes_since=since, limit=100))
-        self.assertEqual(result, {})
-
-    def test_filters_no_metadata(self):
-        """
-        Servers without metadata are not included in the result.
-        """
-        servers = [{'id': i} for i in range(10)]
-        eff = resolve_retry_stubs(get_scaling_group_servers())
-        fake_response = object()
-        body = {'servers': servers}
-        result = resolve_svcreq(eff, (fake_response, body), *self.req)
-        self.assertEqual(result, {})
+        self.now = datetime(2010, 5, 31)
+        self.servers1 = [{'id': 'a', 'a': 'b'}, {'id': 'b', 'b': 'c'}]
+        self.servers2 = [{'id': 'd', 'd': 'e'}]
+        self.freeze = compose(set, map(freeze))
+
+    def _invoke(self):
+        return get_scaling_group_servers(
+            'tid', 'gid', self.now, cache_class=EffectServersCache,
+            all_as_servers=intent_func("all-as"),
+            all_servers=intent_func("alls"))
+
+    def _test_no_cache(self, empty):
+        current = [] if empty else self.servers1
+        sequence = [
+            (("cachegstidgid", False), lambda i: (object(), None)),
+            (("all-as",), lambda i: {} if empty else {"gid": current})]
+        self.assertEqual(perform_sequence(sequence, self._invoke()), current)
+
+    def test_no_cache(self):
+        """
+        If cache is empty then current list of servers are returned
+        """
+        self._test_no_cache(False)
+        self._test_no_cache(True)
+
+    def _test_old_case(self, last_update, as_srvs=True, cur_srvs=True):
+        exp_last_update = datetime(2010, 5, 1)
+        as_servers = self.servers1
```

Comment:
There is no overlap between `as-servers` and `current` - would it make sense to add another server to `current` that has the same ID as `as-servers`, so that we can test that current servers take precedence over the changes-since servers?  If the order of the servers passed to `merge_servers` for the older than 30-day case was reversed, this test wouldn't catch it.